### PR TITLE
#169257427 book an accommodation facility

### PR DESCRIPTION
**Pull request by nshaphan**
#### What does this PR do?
Allow user to book an accommodation

#### Description of Task to be completed?
- [x] book accommodation
- [x] change rooms status
- [x] Endpoint documentation

#### How should this be manually tested?
- Clone the repo and do `npm install` to install dependencies
- Run `npm run migrate:undo` to drop existing database, then `npm run migrate` to create the database 
- Run `npm run watch`
- Run `npm run pretest-travis` and then `npm test` to run tests
- Using postman, Add a new user using `POST  /api/v1/signup`
- Using postman, Book an accommodation using `POST /api/v1/booking`
```
// Sample data
{
  "hotelId": 1,
  "arrivalDate": "2019-12-08",
  "leavingDate": "2019-12-08",
  "rooms": [1]
}
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#169257427](https://www.pivotaltracker.com/story/show/169257427)

#### Screenshots (if appropriate)
image

#### Questions:
N/A

### CODE REVIEW

**Review by gildniy — DISMISSED** [review]
Tested and working. It can be merged.

**Comment by collinewait** [comment]
Resolve conflicts @nshaphan

**Review by RaymondMwaura — APPROVED** [review]
This feature works as expected. I recommend you include seed data in the database with the actual rooms preloaded so that the test runs even better, showing the room as unavailable if it has already been booked.

**Review by RaymondMwaura — CHANGES_REQUESTED** [review]
Resolve conflicts first and then I'll reapprove. I found the functionality to be working as expected.

**Review by RaymondMwaura — DISMISSED** [review]
Great work Shaphan!

**Review by Mastel22 — COMMENTED** [review]
Great work Shaphan. I tested and it works fine however a room can be booked even when it doesn't exist in the database. Maybe you can add a middleware to check if the room exists first then we'll be good to go.

**Review by Mastel22 — COMMENTED** [review]
Please validate the date fields too. Because if you enter a previous date the room is still being booked

File: src/controllers/booking.controller.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    const { userId } = res.locals.user;
+
+    const book = {
+      userId,
+      hotelId,
+      arrivalDate,
+      leavingDate
+    };
+    const pResult = rooms.map(async (roomId) => {
```
Comment (collinewait) [inline src/controllers/booking.controller.js]: It is hard to know the meaning of `pResult`. It can be renamed to show its intent.

Comment (nshaphan) [reply src/controllers/booking.controller.js]: I think `promiseResults` can be better

File: src/controllers/booking.controller.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      });
+      const room = await roomService.changeStatus(roomId, 'reserved');
+
+      return {
+        bookingId: booking.id,
+        room: await db.room.findOne({ where: { id: room } })
+      };
+    });
+
+    const result = await Promise.all(pResult);
```
Comment (collinewait) [inline src/controllers/booking.controller.js]: What does `result` mean? A better name can make code to be easy to read and understand.

File: src/controllers/booking.controller.js (hunk 1 above)
Comment (collinewait) [reply src/controllers/booking.controller.js]: It looks to be an array of promises but of which kind?.  promise`Results` leaves us confused.

**Comment by collinewait** [comment]
Please resolve conflicts.

File: src/services/room.service.js [hunk 3]
```diff
@@ -19,6 +19,23 @@ class RoomService {
...
+   * @returns {Object} room
+   */
+  async changeStatus(roomId, status) {
+    const result = await db.room.update({ status }, {
+      where: {
+        id: roomId
+      }
+    });
+
+    return result;
```
Comment (collinewait) [inline src/services/room.service.js]: Rename `result` to a better name.

File: src/controllers/booking.controller.js (hunk 1 above)
Comment (nshaphan) [reply src/controllers/booking.controller.js]: I have changed to `bookedRooms`

File: src/controllers/booking.controller.js (hunk 2 above)
Comment (nshaphan) [reply src/controllers/booking.controller.js]: changed to `bookedRooms`

File: src/middlewares/roomsAvailability.js [hunk 4]
```diff
@@ -24,8 +49,8 @@ const checkForRooms = async (req, res, next) => {
   });
 
   if (unavailableRooms.count > 0) {
-    const bookedRooms = unavailableRooms.rows.map(room => room.id);
-    return Responses.handleSuccess('409', 'rooms already booked', res, { unAvailableRooms: bookedRooms });
+    const bookedRoom = unavailableRooms.rows.map(room => room.id);
```
Comment (gildniy) [inline src/middlewares/roomsAvailability.js]: I think the `bookedRoom` should be in plural `bookedRooms` as it's an array of many ids.

File: src/middlewares/roomsAvailability.js [hunk 5]
```diff
@@ -24,8 +49,8 @@ const checkForRooms = async (req, res, next) => {
...
+    return Responses.handleSuccess('409', 'rooms already booked', res, { unAvailableRooms: bookedRoom });
```
Comment (gildniy) [inline src/middlewares/roomsAvailability.js]: The response code has to be a number `'409'` to `409`

**Review by gildniy — CHANGES_REQUESTED** [review]
Other than these typos is working great. Please fix them.

File: src/services/room.service.js (hunk 3 above)
Comment (nshaphan) [reply src/services/room.service.js]: changed to `bookedRoom`

File: src/middlewares/roomsAvailability.js (hunk 5 above)
Comment (nshaphan) [reply src/middlewares/roomsAvailability.js]: > Great work Shaphan. I tested and it works fine however a room can be booked even when it doesn't exist in the database. Maybe you can add a middleware to check if the room exists first then we'll be good to go.

Thanks @Mastel22 for finding out this, now solved.

**Review by Mastel22 — APPROVED** [review]
Great job Shaphan. I can see that you added on the checkforRooms middleware a conditional to check if the room exists also that you're checking that the leaving date is greater than the arrival date. I think now this PR is now ready.

**Review by gildniy — APPROVED** [review]
Looks great, tested and working.

**Review by b0nbon1 — APPROVED** [review]
Good job @nshaphan , I like how you handled the edge cases for leaving date and arrival date.

**Review by collinewait — APPROVED** [review]

**Pull request merged**